### PR TITLE
Allow hash to be passed into handleRedirectPromise, reset non-msal after processing

### LINKED ISSUE #2268: Login Redirect fails in Angular 10 - MSAL Browser Sample
## Library
- [ ] `msal@1.x.x` or `@azure/msal@1.x.x`
- [x] `@azure/msal-browser@2.1.0`
- [ ] `@azure/msal-angular@0.x.x`
- [ ] `@azure/msal-angular@1.x.x`
- [ ] `@azure/msal-angularjs@1.x.x`

**Important**: Please fill in your exact version number above, e.g. `msal@1.1.3`.

## Framework
Angular: I am testing the Angular 10 - MSAL Browser Sample

## Description
The Angular 10 - MSAL Browser Sample has been updated to include the ability to do a login redirect rather than a login popup.  However using this login redirect, the post login redirect (in the sample set to `http://localhost:4200`) does not correctly refresh and the Angular Web App (or specifically `MsalService`) is not aware of the successful login.  Calling `this.authService.getAllAccounts().length` in the `AppComponent` returns a length of 0.  After refreshing the Chrome browser, the Web Application is able to process the successful login.  In Safari browser the refresh also fails.

Issue is not present when login is provided with popup.

The sample uses `@azure/msal-browser` version 2.0.2, but I have tried 2.1.0 with the same results.

## Security
- [ ] Is this issue security related?

## Regression
- [ ] Did this behavior work before? 
Version: Not aware

## MSAL Configuration

```js
auth: {
      clientId: '6226576d-37e9-49eb-b201-ec1eeb0029b6',
      redirectUri: 'http://localhost:4200'
}
```

## Reproduction steps
1. clone the `Angular 10 - MSAL Browser` Sample
2. `npm install` and `npm start`
3. visit `http://localhost:4200` and start login redirect by clicking "Login"
4. After providing Azure AD credentials, browser is redirected to `http://localhost:4200`, note that the "Login" button still shows and the "Logout" button is hidden.

Calling the Microsoft Graph in the `/profile` location will trigger a failing acquireToken call.

When I refresh Chrome, calling authService.getAllAccounts() returns the correct value. In Safari the browser refresh kicks off a failing acquireToken loop.

When I use interactionType: `InteractionType.POPUP` in the interceptor config, the post login redirect immediately shows the correctly refreshed Angular web app and the Logout button is showing rather than the Login button.  So, this seems related to the `InteractionType.REDIRECT` configuration only.

## Browsers/Environment
- [x] Chrome
- [ ] Firefox
- [ ] Edge
- [x] Safari
- [ ] IE
- [ ] Other (Please add browser name here)

### PR DESCRIPTION
**Pull request by jasonnutter**
When using hash routing in Angular and enabling `navigateToLoginRequestUrl`, msal will put restore the non-msal hash before the msal hash has been processed, resulting in Angular navigating too early. Furthermore, the Angular router will clear the msal hash before `handleRedirectPromise` has a chance to parse it, so `handleRedirectPromise` can now be given an explicit hash to process. This allows an app (or the MSAL Angular library) to capture the hash before the Angular router clears it, and then process that hash when `handleRedirectPromise` is called.

### CODE REVIEW

File: change/@azure-msal-browser-2020-10-26-18-16-36-handleredirectpromise-hash.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "patch",
+  "comment": "Allow hash to be passed into handleRedirectPromise, reset non-msal after processing",
+  "packageName": "@azure/msal-browser",
+  "email": "janutter@microsoft.com",
+  "dependentChangeType": "patch",
```
Comment (jasonnutter) [inline change/@azure-msal-browser-2020-10-26-18-16-36-handleredirectpromise-hash.json:6]:
```suggestion
  "dependentChangeType": "none",
```

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 2]
```diff
@@ -115,10 +115,10 @@ export abstract class ClientApplication {
      * auth flows.
      * @returns {Promise.<AuthenticationResult | null>} token response or null. If the return value is null, then no auth redirect was detected.
      */
-    async handleRedirectPromise(): Promise<AuthenticationResult | null> {
+    async handleRedirectPromise(hash?: string): Promise<AuthenticationResult | null> {
```
Comment (jasonnutter) [inline lib/msal-browser/src/app/ClientApplication.ts:118]: Should probably update the jsdoc comments to make it clear hash is optional

**Review by tnorling — APPROVED** [review]
Looks good. Can we add a note to the docs that states the hash is optional and should only be provided explicitly if there's a valid reason to?

**Comment by jasonnutter** [comment]
> Looks good. Can we add a note to the docs that states the hash is optional and should only be provided explicitly if there's a valid reason to?

Yep, totally agree.

File: samples/msal-angular-v2-samples/angular10-browser-sample/src/app/msal/msal.service.ts [hunk 3]
```diff
@@ -27,10 +28,18 @@ interface IMsalService {
...
-        @Inject(MSAL_INSTANCE) private msalInstance: IPublicClientApplication
-    ) {}
+        @Inject(MSAL_INSTANCE) private msalInstance: IPublicClientApplication,
+        private location: Location
+    ) {
+        // Cache the code hash before Angular router clears it
+        const hash = this.location.path(true).split('#').pop();
+        if (hash) {
+            this.redirectHash = `#${hash}`;
+        }
```
Comment (jasonnutter) [inline samples/msal-angular-v2-samples/angular10-browser-sample/src/app/msal/msal.service.ts:41]: FYI @jo-arroyo

**Review by pkanher617 — APPROVED** [review]
docs updates should happen with this but seems ok to me.

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 4]
```diff
@@ -113,6 +113,7 @@ export abstract class ClientApplication {
      * Event handler function which allows users to fire events after the PublicClientApplication object
      * has loaded during redirect flows. This should be invoked on all page loads involved in redirect
      * auth flows.
+     * @param hash Hash to process. Defaults to the current value of window.location.hash. Only needs to be provided explicitly if current value does not contain response to be handled.
```
Comment (jo-arroyo) [inline lib/msal-browser/src/app/ClientApplication.ts:116]:
```suggestion
     * @param hash Hash to process. Defaults to the current value of window.location.hash. Only needs to be provided explicitly response to be handled is not contained in the current value.
```

**Review by jo-arroyo — APPROVED** [review]

File: lib/msal-browser/docs/initialization.md [hunk 5]
```diff
@@ -113,6 +113,8 @@ This will also allow you to retrieve tokens on page reload. See the [onPageLoad
 
 It is not recommended to use both interaction types in a single application.
 
+**Note:** `handleRedirectPromise` will optionally accept a hash value to be processed, defaulting to the current value of `window.location.hash`. This parameter only needs to be provided in scenarios where the current value of `window.location.hash` does not contain the redirect response that needs to be processed. For almost all scenarios, applications should not need to provide this parameter explicitly.
```
Comment (pkanher617) [inline lib/msal-browser/docs/initialization.md:116]:
```suggestion
**Note:** `handleRedirectPromise` will optionally accept a hash value to be processed, defaulting to the current value of `window.location.hash`. This parameter only needs to be provided in scenarios where the current value of `window.location.hash` does not contain the redirect response that needs to be processed. **For almost all scenarios, applications should not need to provide this parameter explicitly.**
```

**Review by pkanher617 — APPROVED** [review]
One suggestion, looks good otherwise

**Review by tnorling — APPROVED** [review]

**Pull request merged**